Instance collision

aidandavey opened the pull request:
This PR aims to add collision for the foliage instancer, mentioned in issue #43 

The headlines :

  - **Physics shapes exclusively handled using the PhysicsServer**
  - **Only creates and destroys edge cells**
  - **Provides visualizations via the Rendering Server**
  - **Works for scene files with multiple shapes**
  - **Supports Game and Editor modes**
  - **Allows teleporting**

Some further details:

It will read any CollisionShape3Ds found in the Terrain3DMeshAsset scene file and store their Shape3Ds and transforms.

When terrain collision is rebuilt in Terrain3DCollision it also spawns shapes for the instances using data read from the data/region/instances dictionary and the Shape3D and transforms stored in the MA. 

In Dynamic mode:

1. Build an array of cell positions to build, skipping cells which already exist.

2. Build a dictionary of instances in said cells.

3. Decompose cells outside of the collision radius and store the instances within for recycling.

4. Compare the list of recyclable  instances against the list of instances to build. Decompose unneeded instances to their component shapes, store for recycling.

5. Generate the new instances, preferably:

    A. Recycle an old instance, simply transforming the shapes. Or...
    B. Reuse spare shapes, transform and update base meshes. Or...
    C. Generate brand new shapes.

6. Destroy and remaining shapes and update an internal RID/shape_id map.

Full mode for instance collision was shelved. 

For both Editor and Game modes, the PhysicsServer3D is used. The main reason for this was I kept hitting the 2^18 object limit when stress testing Full mode.

For Editor modes, visualizations are provided via the RenderingServer. 

To test it, add a new Terrain3DMeshAsset and use RockA.tscn as the scene file. Paint some rocks using the foliage instancer and run the scene.

Future plans:

1. Have a body per MA - allow individual layers and other physics settings
2. Manual mode - allow the user to enable and destroy instance collision by cell through the API
3. Roll up of the shape transforms - currently it takes the local transform only
4. Enable update by MA per cell, to improve performance in the editor (if needed)
5. Use hash maps instead of Dictionaries

CavemanIke commented:
> When collision is rebuilt in Terrain3DCollision it also spawns bodies and shapes for the instances using data read from the data/region/instances dictionary and the Shape3D stored in the MA.

If this is creating a body with the shape for each instance, I think an optimization can be had here. If a single body is used and the same shape is added to it multiple times with transforms matching the instances, it will use significantly less memory. Through my tests, adding 250K bodies using the same shape RID results in ~400 megabytes of additional RAM usage. Using a single body, but adding 250K of the same shape to it at different transforms matching the instances was only ~100 megabytes extra RAM usage. Rebuilding shapes on a single body is also extremely fast compared to creating multiple bodies.

aidandavey commented:
Work on only updating edge cells, and spawning body per cell, shape per instance

This will be helpful:

Terrain3DInstancer::_get_cell(const Vector3 &p_global_position, const int p_region_size) will give the cell index for the current region.

so cell_local_pos.xz = instance_pos.xz - cell * cell_size.

TokisanGames commented:
Thanks, I will look when I can. I am a bit behind.

You could add edit the collision document, and document the new functions in the XML (generate it then edit it) if you like.

> Merge branch 'TokisanGames:main' into Instance-collision

Never merge. Only rebase. See the contributing doc.

aidandavey commented:
> Thanks, I will look when I can. I am a bit behind.
> 
> You could add edit the collision document, and document the new functions in the XML (generate it then edit it) if you like.
> 
> > Merge branch 'TokisanGames:main' into Instance-collision
> 
> Never merge. Only rebase. See the contributing doc.
> 
> 

Thanks and no worries, I actually dreamed an improvement last night so I need another day or so to try it out.

I'll also take a look into editing the docs and sorry for the merge.

aidandavey commented:
Planned improvement - prebuild the list of cell_locations to create rather than checking every cell for every mesh asset.

aidandavey commented:
> As the user traverses the instancer cells and they leave behind 10 instances of TreeA and approach 10 instances of TreeA, it would be more efficient to just move the transforms, and store any remaining in an unused cache rather than delete and recreate them every time. The latter is slower, will likely fragment memory, and probably run into any limits faster. The former is what the terrain collision does. It builds a fixed number of shapes at initialization, then never creates nor destroys any until shutdown.
> 
> > Merge branch 'TokisanGames:main' into Instance-collision
> > Revert "Merge branch 'TokisanGames:main' into Instance-collision
> 
> This also isn't what we want. Do an interactive rebase, drop both of these commits. Rebase your PR, and force push.

Yes I saw the fixed number of shapes in the terrain collision which is great when there is a fixed number of shapes of the same type.

I can manage active and inactive shapes for each shape type, but how many to create of each? 

What do you think about creating shapes only when there are no unused ones?

aidandavey committed:
Started instance collision

(cherry picked from commit 8bd8728c5dc3c568aba67b0b2df46c0292caae8c)
(cherry picked from commit f7316bca7c39877cc6bbdbda9336b14777753453)

aidandavey committed:
Build from cell audit and shape recycling

Squashed commit:

[f8062bd] Rename visual -> debug_meshes

[80d7fcb] Fix dissapearing debug_meshes

[aac0b9c] Fix collision is persistant on changing scenes (+2 squashed commit)

Squashed commit:

[3afa186] Update logging message

[620c5fa] Do not update visual instance where none exists (+3 squashed commit)

Squashed commit:

[bdca4d8] Rebuild collision on MMI updates (sync MMIs with instance collision)

[b0a1716] Add consts, remove whitespace

[5a3c11f] add more consts (+3 squashed commit)

Squashed commit:

[303d111] Fix instancer/collision debug shapes race

[493568e] Cleanup terrain_3d_collision.cpp

[1a95405] Fix distance_to always == 0.0 (+3 squashed commit)

Squashed commit:

[db1782a] improve instance collision

[2a1af32] update terrain_3d_collision.cpp

[2ae3a9e] Build from cell audit and shape recycling

(cherry picked from commit 44858e3a0df450d610e8dfd4046208817becda8a)
(cherry picked from commit 23cdbb2c5a1824e21de4c86847497d7efd47d071)

aidandavey commented:
Ok I think I'm done with this for now. Ready for review whenever you get to it

aidandavey commented:
The visuals are not always being updated when the meshes are cleared, needs investigation and fixing

Radivarig commented:
Was this closed by mistake?

almic commented:
I have a suggestion for this. You should make each mesh asset create its own body with its own layer and mask. This way I can do something like raycast a "bullet" and get the body, then ask Terrain3D what mesh asset it was and play an appropriate effect. Having different masks would also be great for things like bushes where I want to have a unique layer to detect if I'm in a bush, without being blocked by it. Looking over the code I think this would be trivial to add, just have a mesh id to body dictionary.

Edit: I'm going to start using this, so I would be more than happy to add to this!

almic committed:
Fix ordering of shape data and body set shape

aidandavey commented:
> Edit: I'm going to start using this, so I would be more than happy to add to this!

Please join and/or chat on the discord server so we can coordinate on this

https://discord.gg/tokisan-games-691957978680786944

TokisanGames commented:
Keep in mind this needs to be merged and tested for a while before pushed in the 1.1 release. It's the last, unfinished major development. If per instance static bodies is complicated, it could go in a separate PR, and one or both might wait until 1.2

almic committed:
Fix region selection for full collision mode

- This would previously load 32x32 cells per region, and
  would attempt to load regions that don't exist if region
  size was smaller than 32x32 cells. This also loads faster
  now as a result.

aidandavey committed:
Improve instance collision (+3 squashed commit)

Squashed commit:

[86d0e7e] Separate instance collision modes and other improvements

[d47e74c] Do not clear debug mesh array before queued updates are processed

[f4e5ab9] Fix distance check XOR error, don't pass native types as pointers and pass in radius as a param

aidandavey commented:
OK I'm finally happy with this

TokisanGames commented on src/terrain_3d_util.cpp:
```diff
@@ -519,6 +519,13 @@ void Terrain3DUtil::benchmark(Terrain3D *p_terrain) {
 	}
 }
 
+Vector2i Terrain3DUtil::get_cell(const Vector3 &p_global_position, const int &p_region_size, const real_t &p_vertex_spacing, const int &p_cell_size) {
```
By reference & isn't needed for the native types.
P_cell_size should be checked for 0, or use the hard coded cell size since it's not changeable. Terrain3DInstancer::CELL_SIZE.
This should be exposed to gdscript.

almic committed:
Convert cell locations to Vector2i, restructure

- Track cell locations as terrain-relative cell positions,
  similar to how region locations are stored
- Restructure instance data dictionary to reduce memory
  usage for transforms
- Optimize some calculations related to cell positions

aidandavey committed:
Fix debug meshes when saving/reloading, update on set vertex spacing (+2 squashed commit)

Squashed commit:

[07f467b] Instance collision set_is_dirty() default value = true

aidandavey committed:
Check if circle overlaps cell in build/desompose logic

aidandavey committed:
Set up one body per MA with layers, masks and physics materials

aidandavey committed:
Fix destroy instance collision

aidandavey committed:
Fix variant cast and rename some things

aidandavey committed:
Fix improper casting
